portafly/src/components/pages/product/CreateProductPage.tsx

```diff
@@ -0,0 +1,197 @@
+import React, {
+  useState,
+  FormEventHandler,
+  FocusEventHandler,
+  useEffect
+} from 'react'
+
+import {
+  PageSection,
+  TextContent,
+  Text,
+  Card,
+  CardBody,
+  Form,
+  FormGroup,
+  TextInput,
+  TextArea,
+  ActionGroup,
+  Button
+} from '@patternfly/react-core'
+import { useTranslation } from 'i18n/useTranslation'
+import { useHistory, Redirect } from 'react-router'
+import { createProduct, NewProduct } from 'dal/products'
+import { useAsync } from 'react-async'
+import { useAlertsContext, useDocumentTitle } from 'components/util'
+import { ValidationException, Validator } from 'utils'
+
+type Validations = Record<string, {
+  validation: 'default' | 'success' | 'error',
+  errors?: string[]
+}>
+
+const CreateProductPage = () => {
+  const { t } = useTranslation('product')
+  useDocumentTitle(t('create.pagetitle'))
+  const { goBack } = useHistory()
+  const { addAlert } = useAlertsContext()
+
+  const [name, setName] = useState('')
+  const [systemName, setSystemName] = useState('')
+  const [description, setDescription] = useState('')
+  const [validations, setValidations] = useState<Validations>({
+    name: { validation: 'default' },
+    system_name: { validation: 'default' },
+    description: { validation: 'default' }
+  })
+
+  const {
+    isPending, error, run, data
+  } = useAsync({ deferFn: createProduct })
+
+  useEffect(() => {
+    if (error) {
+      if (Object.prototype.hasOwnProperty.call(error, 'validationErrors')) {
+        const { validationErrors } = (error as unknown as ValidationException)
+        const newValidations: Validations = {}
+        Object.keys(validationErrors).forEach((id) => {
+          newValidations[id] = {
+            validation: 'error',
+            errors: validationErrors[id]
+          }
+        })
+        setValidations({ ...validations, ...newValidations })
+      } else {
+        addAlert({ id: String(Date.now()), title: error.message, variant: 'danger' })
+      }
+    }
+  }, [error])
+
+  if (data) {
+    const { service } = data as NewProduct
+    return <Redirect to={`/products/${service.id}`} />
+  }
+
+  const isValid = validations.name.validation === 'success' && validations.system_name.validation !== 'error'
+
+  const validator = Validator()
+    .for('name', () => (name.length > 0 ? 'success' : 'error'))
+    // eslint-disable-next-line no-nested-ternary
+    .for('system_name', () => (!systemName ? 'default' : (systemName.length > 0 ? 'success' : 'error')))
+    .for('description', () => (description.length > 0 ? 'success' : 'default'))
```

Comment:
I like it more than the switch case version, good!

I was thinking that maybe we can pass the `Validator` factory a validation rules object like:

```suggestion
  const validator = Validator({
    name: () => (name.length > 0 ? 'success' : 'error'),
    'system_name': () => (!systemName ? 'default' : (systemName.length > 0 ? 'success' : 'error')),
    description: () => (description.length > 0 ? 'success' : 'default'))
  })
   ```

then in the implementation you could assert the case with `Object.prototype.hasOwnProperty` and execute the fn

Reply:
But isn't that how it works already? Only with a sugar coating on top of it https://github.com/3scale/porta/pull/2037/files#diff-9d8e17e1d9797344cd149b6d58dc6377